js/ad_manager_vast.js

```diff
@@ -1875,6 +2127,625 @@ OO.Ads.manager(function(_, $) {
+
+    /**
+     * Get tracking events.
+     * This is only required for VPAID ads
+     * @public
+     * @method Vast#_getTracking
+     * @param {object} parent DOM Element to look for tracking events
+     * @return {array} Array with tracking events and urls
+     */
+    this._getTracking = function(parent) {
+      var node, nodes, tracking, _i, _len;
+      tracking = [];
+      nodes = parent.getElementsByTagName('Tracking');
+      if (!nodes) {
+        return;
+      }
+      for (_i = 0, _len = nodes.length; _i < _len; _i++) {
+        node = nodes[_i];
+        tracking.push({
+          event: node.getAttribute('event'),
+          url: node.textContent
+        });
+      }
+      return tracking;
+    };
+
+    /**
+     * Send error.
+     * This is only required for VPAID ads
+     * @public
+     * @method Vast#_sendError
+     */
+    this._sendError = function() {
+      if (currentAd && currentAd.data) {
+        var error = currentAd.data.error;
+        if (error) {
+          OO.pixelPing(error);
+        }
+      }
+    };
+
+    /**
+     * Send impressions.
+     * This is only required for VPAID ads
+     * @public
+     * @method Vast#_sendImpressions
+     */
+    this._sendImpressions = function() {
+      if (currentAd && currentAd.data) {
+        var impressions = currentAd.data.impressions;
+        _.each(impressions, function(impression) {
+          if (impression && impression.url) {
+            OO.pixelPing(impression.url);
+          }
+        });
+      }
+    };
+
+    /**
+     * Send tracking events.
+     * This is only required for VPAID ads
+     * @public
+     * @method Vast#_sendTracking
+     * @param {string} type Event name to be send
+     */
+    this._sendTracking = function(type) {
+      var ad = prevAd ? prevAd : currentAd;
+      if (ad && ad.data) {
+        var tracking = ad.data.tracking,
+            currentEvent;
+        if (tracking) {
+          currentEvent = _.find(tracking, function(item, index) {
+            return item.event == type
+          });
+
+          if (currentEvent && currentEvent.url) {
+            OO.pixelPing(currentEvent.url);
+          }
+        }
+      }
+    };
+
+    /**
+     * Send click tracking event.
+     * This is only required for VPAID ads
+     * @public
+     * @method Vast#_sendClickTracking
+     * @param {object} ad Ad to look for the tracking url
+     */
+    this._sendClickTracking = function(ad) {
+      if (ad && ad.data && ad.data.videoClickTracking) {
+        var clickTracking = ad.data.videoClickTracking.clickTracking;
+        if (clickTracking){
+          OO.pixelPing(clickTracking);
+        }
+      }
+    };
+
+    /**
+     * Listen and executes events sent by the ad unit
+     * This is only required for VPAID ads
+     * @private
+     * @method Vast#_onAdEvent
+     * @param {string} eventName Name of the event to process
+     */
+    var _onAdEvent = _.bind(function(eventName) {
+      switch(eventName) {
```

Comment:
We should consider try/catches to wrap any vpaidAd API calls. This will ensure that things like tracking don't get lost if the creative has an issue.